### ISSUE TITLE
SQL Server: Derived and aliased fields in Group By cause errors

### ISSUE THREAD

**Issue by rovats** [jOOQ/jOOQ#5588]
If a derived field (eg: `case` statement) with an alias is used in a `group by` clause, the query generated uses the alias in the `group by` clause instead of the definition of the derived field, which is not allowed under SQL Server (this is allowed on MariaDB and possibly other databases).

Eg:
```
Field<String> caseField = DSL.when(F1.isNotNull(), "A").otherwise("B").as("case_field");
dsl.select(caseField, DSL.count())
   .from(TABLE)
   .groupBy(caseField)
   .fetch();
```

This will generate the following, which doesn't work on SQL Server:
```
select case when f1 is not null then 'A' else 'B' end as 'case_field', 
       count(*)
from TABLE
group by case_field
```
The correct query for SQL Server is:

```
select case when f1 is not null then 'A' else 'B' end as 'case_field', 
       count(*)
from TABLE
group by case when f1 is not null then 'A' else 'B' end
```

**Comment by lukaseder** [comment]
Thank you very much for your report, and sorry for the delay. The generated SQL is, in fact, correct, because you're passing exactly this query to jOOQ. The query you wanted to write instead is this one:

```java
Field<String> caseField = DSL.when(F1.isNotNull(), "A").otherwise("B");
dsl.select(caseField.as("case_field"), DSL.count())
   .from(TABLE)
   .groupBy(caseField)
   .fetch();
```

Notice that you also wrote your query this way in SQL.

The reason for jOOQ's behavior is simple. What could the possible intentions be if a field alias (i.e. an ordinary field name) is passed to the group by clause? It can only mean that the user intended the alias to be rendered at that position, not the full expression (because then, the user would pass the expression instead).

I hope this helps,
Lukas

**Comment by rovats** [comment]
Thanks for you reply. 

While I agree with your explanation, the issue here is that some databases allow an alias in the `group by` clause and some do not. So while I am developing against MariaDB, our application also supports MSSQL, and the code that I wrote worked fine on MariaDB but failed later when run on MSSQL.

I'll leave it for you to decide if this difference in behaviour should be handled by jOOQ or not, because otherwise, it gets tricky if developers have to handle these differences in dialect when writing jOOQ code.

**Comment by lukaseder** [comment]
I understand your concern, but let me explain why in this particular case I think it would be unwise to emulate anything to accommodate this MySQL "feature". Assume this:

```sql
SELECT a, COUNT(*) x
FROM t
GROUP BY a, x
```

What does it even mean? Or worse:

```sql
SELECT a, COUNT(*) OVER() x
FROM t
GROUP BY a, x
```

Or this:

```sql
SELECT a AS x, x + SUM(x) -- What will be summed here and why?
FROM (
  SELECT 1 AS a, 2 AS x
) t
GROUP BY x
```

Now check this out:

```sql
SELECT a AS x, 1 AS x, x + SUM(x)
FROM (
  SELECT 1 AS a, 2 AS x
  UNION ALL
  SELECT 2 AS a, 2 AS x
) t
GROUP BY x -- groups by the *last* x in SELECT??
```

What's the meaning of `GROUP BY x` now? `t.x` (because of the `SELECT` alias) or `t.a` (because that's what all the other databases would do)? Now, of course there's always an answer to these questions, but this answer might be extremely complicated. jOOQ keeps it simple. If you reference an aliased column outside of a `SELECT` clause, the alias will be generated. Otherwise, the alias declaration will be generated.

This particular MySQL feature is a big mistake. There are tons of other similar mistakes MySQL made in their early days by not adhering to the SQL standard, and there is the [MySQL "STRICT" mode](https://blog.jooq.org/2014/11/20/use-mysqls-strict-mode-on-all-new-projects/), which I really suggest everyone turns on with new projects.

So, I have these reasons to not go forward with this particular emulation:

1. It was a bad idea by the MySQL people in the first place
2. It will break backwards compatibility (in case someone expects that aliased `x` to be a column from within `t`
3. My workaround is conceptually the same. Even better, I think it makes more sense when you think about it.
4. It would be *very* complicated to implement compared to the little benefit it might provide
5. It would be very unexpected to anyone who is not also using MySQL at the same time (that could be mitigated by adding a `Setting` to enable/disable this feature)
6. Developers will get this wrong only once and then learn SQL "the right way" :)

**Comment by rovats** [comment]
Thanks for the detailed explanation.

I agree, this is a mistake in MySQL along with others like allowing fields in the `select` list that are not part of the `group by` (I noticed this once, but didn't fall for it since I primarily develop on MSSQL).

I wasn't aware of the server modes in MySQL (only started using it a couple of months ago), so thanks for letting me know!

**Comment by lukaseder** [comment]
> I agree, this is a mistake in MySQL along with others like allowing fields in the select list that are not part of the group by (I noticed this once, but didn't fall for it since I primarily develop on MSSQL).

This one is subtle, in fact! The SQL standard allows for putting all *functionally dependent* columns that depend on the `GROUP BY` expression in the `SELECT` clause. It is legal per the SQL standard to write things like:

```sql
SELECT a.first_name, a.last_name, COUNT(*)
FROM actor AS a
JOIN film_actor AS fa USING (actor_id)
JOIN film AS f USING (film_id)
GROUP BY a.actor_id
```

*iff* there is a primary/unique key on `actor.actor_id`, which means that all other columns in `actor` are now *functionally dependent* on `actor_id`.

This behaviour is implemented in PostgreSQL and MySQL to my knowledge (MySQL still does this, when you set it to `STRICT` mode, see: http://dev.mysql.com/doc/refman/5.7/en/sql-mode.html#sqlmode_only_full_group_by).

There's a feature request in jOOQ to emulate this: https://github.com/jOOQ/jOOQ/issues/4725, but to this point, I'm not sure if it's easy to implement...

> I wasn't aware of the server modes in MySQL (only started using it a couple of months ago), so thanks for letting me know!

You're welcome! :)
